### DISCUSSION TITLE
BLE with  esp32_basic_synth

### DISCUSSION THREAD

**Discussion by VorkHeftruck** [marcel-licence/esp32_basic_synth#86] [Q&A] [answered]
Hi Marcel,

First of all thanks for these modules!
I'm trying to get esp32_basic_synth to work with Midi over BLE and i've found some code refering to this in this module. 

#ifdef BLE_MIDI
    Serial.printf("Initialize MIDI over Bluetooth\n");
    BLE_setup();
#endif

Did you already implement BLE? I can't find any related functions in the other modules.
Thanks in advance,
Joost

**Comment by marcel-licence** [comment, marked as answer]
Hi Joost,

you are welcome.
Yes I tried some experiments with BLE. The problem seems that the activation of Bluetooth requires to much processing time and makes it impossible to drive the synthesizer at the same time. If you have a second ESP32 you can setup it as a receiver and just push out the serial data.
I have a little project with three ESP32 (Synth, Bluetooth RX->MIDI, MIDI->Bluetooth Tx) in my queue. It could be an option.

**Comment by VorkHeftruck** [comment]
Hi Marcel,

Thanks for your swift reply. Yes, I was afraid that the Bluetooth  overhead might take too much cpu time.  Still, I thought of giving it a try. Using an additional ESP for communication is feasible although I might go directly from the RPi that acts as a BLE midi server at the moment. I'll give it some thought and maybe some expiriments. Which BLEMidi library did you use?

**Comment by VorkHeftruck** [comment]
Hi Marcel,

Just to keep you updated; I've tried using BLE Midi with your basic synth organ example (wow, your Github is like a candy store; so many great repos) and it's working fine. I'll post in that repo with updates. But for now; I temporarily disabled all the midi logic I could find. I've used https://github.com/lathoub/Arduino-BLE-MIDI as a lib. Ran the Bluetooth process on core0 together with the terminal and haven't found any hiccups yet. And called the noteOn/noteOff directly. I've you want more info please let me know.

**Reply by marcel-licence** [reply]
Hi Joost,

thank you for the updates. I didn't tried the organ yet. The simple organ does not need so much processing power. In my tests I used the basic synth or fm and so on. It looked like that it does not really work in case I just added the include with:
`BLEMIDI_CREATE_DEFAULT_INSTANCE()` without any additional calls.

I was also using: https://github.com/lathoub/Arduino-BLE-MIDI
Which ESP32 board library did you use? I discovered changes with memory management etc. causing all my projects to fail when upgrading from 1.6.0 to 2.#

**Comment by VorkHeftruck** [comment]
Hi Marcel,

Ok, thanks. I'm still using v1.0.6 of the ESP boardmanager. If it's easy to roll back I could check v2?

**Reply by marcel-licence** [reply]
Hi Joost,

I am not sure if the roll back is easy. I've installed both. There is a way to have both at the same time. You can just check out the other board to ..Documents\Arduino\hardware\espressif\esp32_2_0_2
It will create some errors because libraries expect to see "esp32" as folder name instead of another name.

I will try the organ with 1.0.6. Maybe it might work, that would be great in that case I can also make a short video of it.

**Reply by VorkHeftruck** [reply]
Thanks for the board manager tip.
If there's anything I can do to help out or test, please let me know. I really like your videos!

**Reply by marcel-licence** [reply]
Thanks a lot!
I got it also running. It would be nice to know if other synths would support this feature. I tried to setup the more CPU intensive "pro" variant of the organ. It failed.
I will also give the latest ESP32 library a try.

In terms of latency ... you will loose all the fun to play this instrument because it does does not feel connected tight to the keyboard input. But all in all a nice experiment

**Reply by VorkHeftruck** [reply]
Thanks for the info!
I've tried giving esp32_basic_synth a go with the same procedure as I did with basic organ. No go just yet. I'm getting panics on core 0, somewhere in the main loop. Will debug some more but will probably give it a rest and try your synth library directly in a more controlled and simple code setup.
